docs/docs/adrs/adr-011-improving-test-confidence.md

```diff
@@ -0,0 +1,208 @@
+
+## Integration testing
+With integration testing we **test the multi-module interactions** while isolating them from the remainder of the system.
+Integration tests can uncover bugs that are often missed by unit tests.
+
+It is very difficult to gauge the actual test coverage imparted by integration tests and the available tooling is limited.
+In interchain-security we employ the `ibc-go/testing` framework to test interactions in-memory.
+
+At present, integration testing does not involve the consensus layer - it is only concerned with application level state and logic.
+
+## End-to-end testing
+In our context end-to-end testing comprises of tests that use the actual application binaries in an isolated environment (e.g. docker container). During test execution the inputs are meant to simulate actual user interaction, either by submitting transactions/queries using the command line or using gRPC/REST APIs and checking for state changes after an action has been performed. With this testing strategy we also include the consensus layer in all of our runs. This is the closest we can get to testing user interactions without starting a full testnet.
+
+End-to-end testing strategies vary between different teams and projects and we strive to unify our approach to the best of our ability (at least for ICS and gaia).
+
+The available tooling does not give us significant (or relevant) line of code coverage information since most of the tools are geared towards analyzing unit tests and simple code branch evaluation.
+
+We aim to adapt our best practices by learning from other similar systems and projects such as cosmos-sdk, ibc-go and CometBFT.
+
+
+## Decision
+
+### 1. Connect specifications to code and tooling
+Oftentimes, specifications are disconnected from the development and QA processes. This gives rise to problems where the specification does not reflect the actual state of the system and vice-versa. 
+Usually usually specifications are just text files that are rarely used and go unmaintained after a while, resulting in consistency issues and misleading instructions/expectations about system behavior.
+
+#### Decision context and hypothesis
+Specifications written in a dedicated and executable specification language are easier to maintain than the ones written entirely entirely in text.
+Additionally, we can create models based on the specification OR make the model equivalent to a specification.
+
+Models do not care about the intricacies of implementation and neither do specifications. Since both models and specifications care about concisely and accurately describing a system (such as a finite state machine), we see a benefit of adding model based tools (such as [quint](https://github.com/informalsystems/quint)) to our testing and development workflows.
```

Comment:
Quint does seem cool and all, but it seems from @p-offtermatt's experience that it doesn't suit all our needs in its current form. 

We should consider a tool like https://github.com/flyingmutant/rapid to define our model, and generate high level test cases. Sure, we wouldn't be writing things in a TLA-esq syntax. But rapid might provide the usefulness we need in the short to medium term.

You could conceivably write an executable spec in the language of the rapid framework, arguably in a more readable form to layman like me who don't know TLA+.  

I favor the simplicity of writing everything (both the model and driver) in golang, avoiding the need for specialized knowledge in an experimental specification language, and avoiding a test trace interpreter.

My question here is, what does Quint offer that rapid does not? I'm failing to see the value props of using Quint in its current form besides the fact that it's cool to write your spec in a TLA-esq syntax and we want to support Informal projects

Reply:
Good point Shawn, I also considered using Rapid.

My feeling is that it's probably very hard to get a specification of reasonable abstraction level with Rapid,
potentially harder than with Quint. This is just an educated guess, but part of the complexity is just that the protocol is inherently complex.

My current intuition is that if we are just interested in good test coverage, doing property-based testing with Rapid is probably better in terms of cost/benefit, but this wouldn't yield executable specs. 

If we are interested in executable specs, Quint still seems reasonable to me, but I don't have a good feeling on how to get better data on this.

Reply:
Good points, here's some followup questions to consider

> My feeling is that it's probably very hard to get a specification of reasonable abstraction level with Rapid

Why is this the case? What makes a protocol easier to describe/abstract with Quint compared to Rapid, or any language/framework for that matter? Does Rapid have any functionality/advantages that Quint does not? 

Maybe we can try modeling just a subprotocol of replicated security with both Quint and Rapid, and compare side by side how challenging it is to model with each.

Regarding the challenge of creating a succinct model: The particular language or framework we use matters. But generally well written and encapsulated software matters just as much imo

Reply:
Yeah, good points here Shawn. The working hypothesis is that for high-level specifications where implementation-level details are abstracted away, a language that purposely abstracts a lot of these details away is easier to use than a language that needs more fine-grained management of details like memory, concurrency, ... for performance reasons. We will see how it turns out, but I agree that giving both a try could be a good idea.